### PR TITLE
Segment attribute

### PR DESCRIPTION
**Pull request by tspring**
Change allows adding custom attributes to the tracer via a Segment.addCustomAttribute API. It assumes all instances of AbstractTracer are able to send attributes outside of a Transaction.

### CODE REVIEW

**Comment by tspring** [comment]
Screen Shot 2020-09-19 at 8 34 45 PM

File: agent-bridge/src/main/java/com/newrelic/agent/bridge/NoOpSegment.java [hunk 1]
```diff
@@ -67,4 +69,20 @@ public void end(){
...
+    @Override
+    public void addCustomAttribute(String key, Number value) {
+    }
+
+    @Override
+    public void addCustomAttribute(String key, String value) {
+    }
+
+    @Override
+    public void addCustomAttribute(String key, boolean value) {
```
Comment (gfuller1) [inline agent-bridge/src/main/java/com/newrelic/agent/bridge/NoOpSegment.java:81]: Boolean

File: newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java [hunk 2]
```diff
@@ -187,7 +188,10 @@ public void testEndSameThread() throws InterruptedException {
         Assert.assertNotNull(root);
         Assert.assertNotNull(root.getTransactionActivity().getTransaction());
         Thread.sleep(1);
-        Segment segment = root.getTransactionActivity().getTransaction().startSegment(MetricNames.CUSTOM, "Custom Sync Segment");
+        Segment segment = root.getTransactionActivity()
+                .getTransaction()
+                .startSegment(MetricNames.CUSTOM, "Custom Sync Segment");
+        segment.addCustomAttribute("redbeans", "rice");
```
Comment (gfuller1) [inline newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java:194]: does this need to be added here?

File: newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java [hunk 3]
```diff
@@ -1140,6 +1210,65 @@ public void setHeader(String name, String value) {
         assertEquals(segmentSpanEvent.getGuid(), secondRootTracerSpanEvent.getParentId());
     }
 
+    @Test
+    public void testSegmentAddCustomAttributeSync() {
+        Transaction.clearTransaction();
+        Tracer rootTracer = makeTransaction();
```
Comment (gfuller1) [inline newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java:1216]: lets finish this tracer for better cleanup

File: newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java [hunk 4]
```diff
@@ -1140,6 +1210,65 @@ public void setHeader(String name, String value) {
...
+        Transaction tx = rootTracer.getTransactionActivity().getTransaction();
+        Segment segment = tx.startSegment("custom", "segment");
```
Comment (gfuller1) [inline newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java:1218]: lets end the segment at the end of the test before the assertions

File: newrelic-api/src/main/java/com/newrelic/api/agent/AttributeHolder.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package com.newrelic.api.agent;
+
+import java.util.Map;
+
+
+public interface AttributeHolder {
+
+    /**
+     * Adds/Replaces a numerical attribute on the current tracer.
```
Comment (gfuller1) [inline newrelic-api/src/main/java/com/newrelic/api/agent/AttributeHolder.java:9]: remove "on the current tracer" for this one and below

File: newrelic-api/src/main/java/com/newrelic/api/agent/Segment.java [hunk 6]
```diff
@@ -7,6 +7,7 @@
 
 package com.newrelic.api.agent;
 
+
```
Comment (gfuller1) [inline newrelic-api/src/main/java/com/newrelic/api/agent/Segment.java:10]: remove

File: newrelic-api/src/main/java/com/newrelic/api/agent/Segment.java [hunk 7]
```diff
@@ -92,4 +93,6 @@
      * @since 5.4.0
      */
     void endAsync();
+
+
```
Comment (gfuller1) [inline newrelic-api/src/main/java/com/newrelic/api/agent/Segment.java:97]: remove these 2

File: newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java (hunk 3 above)
Comment (tspring) [reply newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java:1216]: It looks like it's taken care of in the @After method

**Review by gfuller1 — CHANGES_REQUESTED** [review]
small stuff

**Comment by breedx-nr** [comment]
Please confirm that this resolves #47

File: newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java (hunk 3 above)
Comment (gfuller1) [reply newrelic-agent/src/test/java/com/newrelic/agent/SegmentTest.java:1215]: I don't see the tracer or segment explicitly cleaned up in the @After method. The transaction is cleared but I think it's still good practice to finish the objects in the test.

**Review by gfuller1 — APPROVED** [review]

**Pull request merged**